lib/block-supports/duotone.php

```diff
@@ -0,0 +1,127 @@
+<?php
+/**
+ * Duotone block support flag.
+ *
+ * @package gutenberg
+ */
+
+/**
+ * Registers the style and colors block attributes for block types that support it.
+ *
+ * @param WP_Block_Type $block_type Block Type.
+ */
+function gutenberg_register_duotone_support( $block_type ) {
+	$has_duotone_support = false;
+	if ( property_exists( $block_type, 'supports' ) ) {
+		$has_duotone_support = gutenberg_experimental_get( $block_type->supports, array( 'duotone' ), false );
+	}
+
+	if ( $has_duotone_support ) {
+		if ( ! $block_type->attributes ) {
+			$block_type->attributes = array();
+		}
+
+		if ( ! array_key_exists( 'duotone', $block_type->attributes ) ) {
+			$block_type->attributes['duotone'] = array(
+				'type' => 'object',
+			);
+		}
+	}
+}
+
+
+/**
+ * Add CSS classes and inline styles for colors to the incoming attributes array.
+ * This will be applied to the block markup in the front-end.
+ *
+ * @param  WP_Block_Type $block_type       Block type.
+ * @param  array         $block_attributes Block attributes.
+ *
+ * @return array Colors CSS classes and inline styles.
+ */
+function gutenberg_apply_duotone_support( $block_type, $block_attributes ) {
+	$attributes          = array();
+	$has_duotone_support = false;
+	if ( property_exists( $block_type, 'supports' ) ) {
+		$has_duotone_support = gutenberg_experimental_get( $block_type->supports, array( 'duotone' ), false );
+	}
+	if ( $has_duotone_support ) {
+		$has_duotone_attribute = array_key_exists( 'duotone', $block_attributes );
+
+		if ( $has_duotone_attribute ) {
+			$attributes['class'] = $block_attributes['duotone']['id'];
+		}
+	}
+
+	return $attributes;
+}
+
+/**
+ * Render out the duotone stylesheet and SVG.
+ *
+ * @param  WP_Block_Type $block_type       Block type.
+ * @param  array         $block_attributes Block attributes.
+ * @param  string        $block_content    Rendered block content.
+ *
+ * @return string filtered block content.
+ */
+function gutenberg_render_duotone_support( $block_type, $block_attributes, $block_content ) {
+	$duotone_support = false;
+	if ( property_exists( $block_type, 'supports' ) ) {
+		$duotone_support = gutenberg_experimental_get( $block_type->supports, array( 'duotone' ), false );
+	}
+
+	$has_duotone_attribute = array_key_exists( 'duotone', $block_attributes );
+
+	if (
+		! $duotone_support ||
+		! $has_duotone_attribute
+	) {
+		return $block_content;
+	}
+
+	$duotone_id     = $block_attributes['duotone']['id'];
+	$duotone_values = $block_attributes['duotone']['values'];
+
+	// Adding the block class as to not affect other blocks.
+	$block_class = gutenberg_get_block_default_classname( $block_type->name );
+	$scope       = '.' . $block_class . '.' . $duotone_id;
```

Comment:
I think I'm starting to become more convinced  that we should (at least for the initial version) start with a unique generated id like I'm  doing in https://github.com/WordPress/gutenberg/pull/29335